core/adapters/http.go

```diff
@@ -169,54 +170,90 @@ func sendRequest(input models.RunInput, request *http.Request, config HTTPReques
 	if err != nil {
 		return models.NewRunOutputError(err)
 	}
 
-	defer response.Body.Close()
-
-	source := newMaxBytesReader(response.Body, config.sizeLimit)
-	bytes, err := ioutil.ReadAll(source)
-	if err != nil {
-		return models.NewRunOutputError(err)
-	}
-
-	responseBody := string(bytes)
-	if response.StatusCode >= 400 {
+	// This is either a client error caused on our end or a server error that persists even after retrying.
+	// Either way, there is no way for us to complete the run with a result.
+	if statusCode >= 400 {
 		return models.NewRunOutputError(errors.New(responseBody))
 	}
 
 	return models.NewRunOutputCompleteWithResult(responseBody)
 }
 
+// withRetry executes the http request in a retry. Timeout is controlled with a context
+// Retry occurs if the request timeout, or there is any kind of connection or transport-layer error
+// Retry also occurs on remote server 5xx errors
 func withRetry(
 	client *http.Client,
 	originalRequest *http.Request,
 	config HTTPRequestConfig,
-) (*http.Response, error) {
-	var response *http.Response
-	err := retry.Do(
+) (responseBody string, statusCode int, err error) {
+	err = retry.Do(
 		func() error {
 			ctx, cancel := context.WithTimeout(context.Background(), config.timeout)
 			defer cancel()
 			requestWithTimeout := originalRequest.Clone(ctx)
 
+			start := time.Now()
+
 			r, err := client.Do(requestWithTimeout)
 			if err != nil {
 				return err
 			}
-			response = r
+			defer r.Body.Close()
+			statusCode = r.StatusCode
+			elapsed := time.Since(start)
+			logger.Debugw(fmt.Sprintf("http adapter got %v in %s", statusCode, elapsed), "statusCode", statusCode, "timeElapsedSeconds", elapsed)
+
+			source := newMaxBytesReader(r.Body, config.sizeLimit)
+			bytes, err := ioutil.ReadAll(source)
+			if err != nil {
+				logger.Errorf("http adapter error reading body: %v", err.Error())
+				return err
+			}
+			elapsed = time.Since(start)
+			logger.Debugw(fmt.Sprintf("http adapter finished after %s", elapsed), "statusCode", statusCode, "timeElapsedSeconds", elapsed)
+
+			responseBody = string(bytes)
+
+			// Retry on 5xx since this might give a different result
+			if 500 <= r.StatusCode && r.StatusCode < 600 {
+				return &ErrRemoteServer{responseBody, statusCode}
+			}
+
 			return nil
 		},
 		retry.Attempts(config.maxAttempts),
+		retry.RetryIf(func(err error) bool {
+			switch err.(type) {
+			// There is no point in retrying a request if the response was
+			// too large since it's likely that all retries will suffer the
+			// same problem
+			case *ErrHTTPResponseTooLarge:
+				return false
+			default:
+				return true
+			}
+		}),
+		retry.OnRetry(func(n uint, err error) {
+			logger.Debugw("http adapter error, will retry", "error", err.Error(), "attempt", n, "timeout", config.timeout)
+		}),
 	)
 
-	if err != nil {
-		return nil, err
-	}
-	return response, nil
+	return responseBody, statusCode, err
+}
+
+type ErrRemoteServer struct {
```

Comment:
Please see https://github.com/golang/go/wiki/Errors#naming for naming convention